src/api/venue.ts

```diff
@@ -34,3 +44,71 @@ export const setVenueLiveStatus = async ({
     })
     .finally(onFinish);
 };
+
+// TODO: return an array of all venueId's we find on the full way up to the sovereignVenue
+export const fetchSovereignVenueId = async (
+  venueId: string,
+  previouslyCheckedVenueIds: readonly string[] = []
+): Promise<string> => {
+  const venue = await fetchVenue(venueId);
+
+  if (!venue) throw new Error(`The '${venueId}' venue doesn't exist`);
+
+  if (!venue.parentId) return venue.id;
+
+  if (previouslyCheckedVenueIds.includes(venueId))
+    throw new Error(
+      `Circular reference detected. '${venueId}' has already been checked`
+    );
+
+  return fetchSovereignVenueId(venue.parentId, [
+    ...previouslyCheckedVenueIds,
+    venueId,
+  ]);
+};
+
+export const fetchVenue = async (
+  venueId: string
+): Promise<WithId<AnyVenue> | undefined> => {
+  const venueDoc = await getVenueRef(venueId).get();
+
+  // TODO: Use proper data validation + firestore model converters to set this type rather than just forcing it with 'as'
+  //  see .withConverter(soundConfigConverter) in fetchSoundConfigs in src/api/sounds.ts as an example
+  const venue: AnyVenue | undefined = venueDoc.data() as AnyVenue | undefined;
+
+  if (!venue) return undefined;
+
+  return withId(venue, venueId);
+};
+
+export const fetchChildVenues = async (
+  venueId: string
+): Promise<WithId<AnyVenue>[]> => {
+  const childVenuesSnapshot = await getVenueCollectionRef()
+    .where("parentId", "==", venueId)
+    // .withConverter()
+    .get();
+
+  if (childVenuesSnapshot.empty) return [];
+
+  // TODO: Use proper data validation + firestore model converters to set this type rather than just forcing it with 'as'
+  //  see .withConverter(soundConfigConverter) in fetchSoundConfigs in src/api/sounds.ts as an example
+  return childVenuesSnapshot.docs
+    .filter((docSnapshot) => docSnapshot.exists)
+    .map((docSnapshot) =>
+      withId(docSnapshot.data() as AnyVenue, docSnapshot.id)
+    );
+};
+
+// TODO: implement fetchRelatedVenues
+export const fetchRelatedVenues = async (
+  venueId: string
+): Promise<WithId<AnyVenue>[]> => {
+  // TODO: find the sovereignVenueId
+  // TODO: fetch the sovereignVenue (note if fetchSovereignVenueId returned the venue and not just the id, we wouldn't need this extra step)
+  // TODO: fetch the sovereignVenue's children
+  // TODO: recursively fetch their children, etc till we have all of them
+  // TODO: return the final array
+
+  return [];
+};
```

Comment:
> TODO: implement fetchRelatedVenues

Plan:

> TODO: find the `sovereignVenueId`
> TODO: fetch the `sovereignVenue` (note if `fetchSovereignVenueId` returned the venue and not just the `id`, we wouldn't need this extra step)
> TODO: fetch the `sovereignVenue`'s children
> TODO: recursively fetch their children, etc till we have all of them
> TODO: return the final array

Reply:
> TODO: return the final array

Do we want to do this as an array? Or do we want to maintain some concept of the hierarchy of the data?

If the latter, then we would probably want to use some form of `VenueNode` type concept similar to:

```typescript
export interface VenueNode {
  parentVenue?: WithId<AnyVenue>;
  selfVenue: WithId<AnyVenue>;
  childVenues: WithId<AnyVenue>[];
}
```

This would allow us to neatly contain all of the relevant structure/layout details in a probably efficient manner. The efficiency of this probably depends on the size of the venue objects.. I believe it would definitely be efficient if we were just storing their IDs rather than the whole objects?

Reply:
> TODO: recursively fetch their children, etc till we have all of them

For this, we can probably use our 'chunked array-in query' pattern to fetch multiple venues at the same time, rather than issuing a request for every child `venueId`.

Reply:
> TODO: find the `sovereignVenueId`
> TODO: fetch the `sovereignVenue` (note if `fetchSovereignVenueId` returned the venue and not just the id, we wouldn't need this extra step)

It would probably be better to refactor `fetchSovereignVenueId` to be `fetchSovereignVenue` (and then we can have a version of it that gets just the `id` if we want/need)

Reply:
> Do we want to do this as an array? Or do we want to maintain some concept of the hierarchy of the data?
> 
> If the latter, then we would probably want to use some form of VenueNode type concept similar to:

Given the possibility of having grandchildren/grandgrandchildren, this concept seems to be impossible to implement?

Reply:
Which concept? An array, or a hierarchy?

Both would be able to support any level of nested data?

Reply:
> Which concept? An array, or a hierarchy of nodes?

I meant that if you would want to return children, grandchildren .... etc, it would not be possible because you don't know how many layers are there?

But if you call all of those `childVenues`, it's fine